warp-tls/Network/Wai/Handler/WarpTLS.hs

```diff
@@ -256,11 +259,34 @@ tlsSettingsChainMemory
     -> [S.ByteString] -- ^ Chain certificate bytes
     -> S.ByteString -- ^ Key bytes
     -> TLSSettings
-tlsSettingsChainMemory cert chainCerts key = defaultTlsSettings
-    { certMemory = Just cert
-    , chainCertsMemory = chainCerts
-    , keyMemory = Just key
-    }
+tlsSettingsChainMemory cert chainCerts key = defaultTlsSettings { 
+    certSettings = CertFromMemory cert chainCerts key
+  }
+
+-- | A smart constructor for 'TLSSettings', but uses references to in-memory
+-- representations of the certificate and key based on 'defaultTlsSettings'.
+--
+-- @since 3.3.0
+tlsSettingsRef 
+    :: I.IORef S.ByteString -- ^ Reference to certificate bytes
+    -> I.IORef (S.ByteString) -- ^ Reference to key bytes 
+    -> TLSSettings 
```

Comment:
I would have argued strongly against providing this function. The most widely used CA on the planet is now "Let's Encrypt" and it along with almost all the other CAs does not sign the leaf certificate directly with a trusted root CA. Instead pretty much all certificates that chain up to one of the usual WebPKI CAs are signed by an _intermediate_ CA.

When you have a certificate signed by an intermediate CA, while various browsers are willing to pay the cost of dynamically locating intermediate CA certs from various certificate extensions, and may have an intermediate cert cache, other software tools (e.g. the ubiquitous `curl`, etc.) are not, and the TLS specifications DO NOT make intermediate certificates optional. They are a required part of the server's certificate chain.

Here the caller can always provide an empty chain if that's what they really want to do, but there's no need to encourage such negligence by providing an API entry point that can only do the wrong thing.  The more general entry point below is the only one that should have been offered.